deprecate TLS_DH_* (and TLS_DHE_* in TLS 1.2) ... (#290)
yaronf/I-D: BCP195bis/draft-ietf-uta-rfc7525bis.md

               with the string "TLS_RSA_WITH_*", have several drawbacks, especially
               the fact that they do not support forward secrecy.
            
* Implementations SHOULD NOT negotiate cipher suites based on
               non-ephemeral (static) finite-field Diffie-Hellman key agreement.

  Rationale: These cipher suites, which have assigned values prefixed by "TLS_DH_*", have several drawbacks, especially
               the fact that they do not support forward secrecy.

* Implementations MUST support and prefer to negotiate cipher suites 

>                offering forward secrecy.  However, given the known fragility
>                of the construction (see {{RACCOON}}) and the limitation around
>                negotiation (including using {{?RFC7919}}, which has seen very
>                limited uptake), TLS 1.2 implementations SHOULD NOT negotiate
>                cipher suites based on ephemeral finite-field Diffie-Hellman key
>                agreement (i.e., "TLS_DHE_*" suites).

Author: ```suggestion
               offering forward secrecy.  However, TLS 1.2 implementations SHOULD NOT negotiate
               cipher suites based on ephemeral finite-field Diffie-Hellman key
               agreement (i.e., "TLS_DHE_*" suites), given the known fragility
               of the construction (see {{RACCOON}}) and the limitation around
               negotiation. including using {{?RFC7919}}, which has seen very
               limited uptake.
```

               with the string "TLS_RSA_WITH_*", have several drawbacks, especially
               the fact that they do not support forward secrecy.
            
* Implementations SHOULD NOT negotiate cipher suites based on

>                non-ephemeral finite-field Diffie-Hellman key agreement.

Reviewer: ```suggestion
               non-ephemeral (static) finite-field Diffie-Hellman key agreement.
```

>   Rationale: These cipher suites, which have assigned values starting
>                with the string "TLS_DH_*", have several drawbacks, especially

Reviewer: ```suggestion
  Rationale: These cipher suites, which have assigned values prefixed by "TLS_DH_*", have several drawbacks, especially
```

               the fact that they do not support forward secrecy.

* Implementations MUST support and prefer to negotiate cipher suites 
               offering forward secrecy, such as those in the Ephemeral 
               Diffie-Hellman and Elliptic Curve Ephemeral Diffie-Hellman ("DHE" 
               and "ECDHE") families.  However, given the known limitations
               around both construction (e.g., {{RACCOON}}) and negotiation
               (including using {{?RFC7919}}), TLS 1.2 implementations SHOULD
               NOT negotiate cipher suites based on ephemeral finite-field

>                Diffie-Hellman key agreement (i.e., "TLS_DHE_*" suites).

Reviewer: Should we clarify what we don't like about 7919: is it still vulnerable? Is it just not well deployed?

Author: explicitly state the reasons why it is not widely implemented / not usable in 1.2

Author: See 5f5c461

>                agreement (i.e., "TLS_DHE_*" suites).

Reviewer: This bullet has become confusing. I suggest to reword and remove DHE from the MUST sentence at the top.

Author: Done in 66c60e0

Clients are of course free to offer stronger cipher suites, e.g., using AES-256; when they do, the server SHOULD prefer the stronger cipher suite unless there are compelling reasons (e.g., seriously degraded performance) to choose otherwise.

> This document does not change the mandatory-to-implement (MTI) TLS cipher suite(s) prescribed by TLS. To maximize interoperability, RFC 5246 mandates implementation of the TLS_RSA_WITH_AES_128_CBC_SHA cipher suite, which is significantly weaker than the cipher suites recommended here. (The GCM mode does not suffer from the same weakness, caused by the order of MAC-then-Encrypt in TLS {{Krawczyk2001}}, since it uses an AEAD mode of operation.) Implementers should consider the interoperability gain against the loss in security when deploying the TLS_RSA_WITH_AES_128_CBC_SHA cipher suite. Other application protocols specify other cipher suites as MTI.

Reviewer: Maybe we should bite the bullet and declare `TLS_RSA_WITH_AES_128_CBC_SHA` dead. I think we don't even need to "update" 5246 because it's already obsolete...

This is out of scope for this PR though.

Author: This is issue #292 now

Note that some profiles of TLS 1.2 use different cipher suites. For example, {{?RFC6460}} defines a profile that uses the TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 and TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 cipher suites.

> {{!RFC8422}} allows clients and servers to negotiate ECDH parameters (curves).  Both clients and servers SHOULD include the "Supported Elliptic Curves" extension {{!RFC8422}}.  Clients and servers SHOULD support the NIST P-256 (secp256r1) curve {{!RFC8422}} and X25519 {{!RFC7748}}. For backwards compatibility purposes, clients MAY send an ec_point_formats extension with a single element, "uncompressed".

Reviewer: Why did you change the SHOULD on ec_point_formats to MAY?

Reviewer 2: We could say "but this is not strictly necessary because "uncompressed" is the only defined value for this extension".

Author: great, will add.

Author: I have reworded the paragraph in 96994b9

> {{!RFC8422}} allows clients and servers to negotiate ECDH parameters (curves).  Both clients and servers SHOULD include the "Supported Elliptic Curves" extension {{!RFC8422}}.  Clients and servers SHOULD support the NIST P-256 (secp256r1) {{!RFC8422}} and X25519 (x25519) {{!RFC7748}} curves.  Note that {{!RFC8422}} deprecates all but the uncompressed point format.  Therefore, if the client sends an ec_point_formats extension, the ECPointFormatList MUST contain a single element, "uncompressed".

Reviewer: See new discussion on CFRG, https://mailarchive.ietf.org/arch/msg/cfrg/Ev8hgyojKeObXMZ7SF2m3_yekMo/

I believe it only applies to EdDSA certificates/signatures, not to ECDH, correct?

We should also clarify this point here, I think.

Author: > See new discussion on CFRG, https://mailarchive.ietf.org/arch/msg/cfrg/Ev8hgyojKeObXMZ7SF2m3_yekMo/

Thanks for the pointer.

> I believe it only applies to EdDSA certificates/signatures, not to ECDH, correct?

Yes, it's correct.

> We should also clarify this point here, I think.

We chose to not recommend ECDSA ciphersuites -- although [it looks like they're not so unpopular](https://notary.icsi.berkeley.edu/#certificate-public-key-details) -- so I don't know exactly where to anchor a clarification.

Reviewer: Hmmm, the ICSI reference is quite old actually, but here's [newer information](https://www.f5.com/labs/articles/threat-intelligence/the-2021-tls-telemetry-report), "The move to elliptic curve cryptography is slow but steady, with 25 percent of certificates now signed with the Elliptic Curve Digital Signature Algorithm (ECDSA)..." - I think we need to reopen that discussion.

Author: See #218

Forward secrecy ensures in such cases that it is not feasible for an attacker to determine the session keys even if the attacker has obtained the long-term keys some time after the conversation. It also protects against an attacker who is in possession of the long-term keys but remains passive during the conversation.

> Forward secrecy is generally achieved by using the Diffie-Hellman scheme to derive session keys. The Diffie-Hellman scheme has both parties maintain private secrets and send parameters over the network as modular powers over certain cyclic groups. The properties of the so-called Discrete Logarithm Problem (DLP) allow the parties to derive the session keys without an eavesdropper being able to do so. There is currently no known attack against DLP if sufficiently large parameters are chosen. A variant of the Diffie-Hellman scheme uses elliptic curves instead of the originally proposed modular arithmetic. Given the current state of the art, elliptic-curve Diffie-Hellman appear to be more efficient, permit shorter key lengths, and allow less freedom for implementation errors than finite-field Diffie–Hellman.

Reviewer: ```suggestion
Forward secrecy is generally achieved by using the Diffie-Hellman scheme to derive session keys. The Diffie-Hellman scheme has both parties maintain private secrets and send parameters over the network as modular powers over certain cyclic groups. The properties of the so-called Discrete Logarithm Problem (DLP) allow the parties to derive the session keys without an eavesdropper being able to do so. There is currently no known attack against DLP if sufficiently large parameters are chosen. A variant of the Diffie-Hellman scheme uses elliptic curves instead of the originally proposed modular arithmetic. Given the current state of the art, elliptic-curve Diffie-Hellman appears to be more efficient, permits shorter key lengths, and allows less freedom for implementation errors than finite-field Diffie–Hellman.
```
